File: lib/core/models/task.dart
enum RepeatType { none, hourly, daily, weekly, monthly }

class Task {
  final String id;
  final String userId;
  final String? categoryId;
  final String categoryName;
  final String title;
  final DateTime? deadline;
  final int? reminderMinutes;
  final RepeatType repeatType;
  final int repeatInterval;
  final DateTime? repeatEndDate;
  final bool isCompleted;
  final DateTime? completedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Task({
    required this.id,
    required this.userId,
    this.categoryId,
    this.categoryName = 'Kerja',
    required this.title,
    this.deadline,
    this.reminderMinutes,
    this.repeatType = RepeatType.none,
    this.repeatInterval = 1,
    this.repeatEndDate,
    this.isCompleted = false,
    this.completedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get hasReminder => reminderMinutes != null;
  bool get hasRepeat => repeatType != RepeatType.none;
  bool get hasDeadline => deadline != null;

  String get repeatTypeString {
    switch (repeatType) {
      case RepeatType.hourly:
        return 'Jam';
      case RepeatType.daily:
        return 'Harian';
      case RepeatType.weekly:
        return 'Mingguan';
      case RepeatType.monthly:
        return 'Bulanan';
      default:
        return 'Tidak';
    }
  }

  String get reminderString {
    if (reminderMinutes == null) return 'Tidak ada';
    return '$reminderMinutes menit sebelumnya';
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      categoryId: json['category_id'] as String?,
      categoryName: json['category_name'] as String? ?? 'Kerja',
      title: json['title'] as String,
      deadline: json['deadline'] != null
          ? DateTime.parse(json['deadline'] as String)
          : null,
      reminderMinutes: json['reminder_minutes'] as int?,
      repeatType: _parseRepeatType(json['repeat_type'] as String?),
      repeatInterval: json['repeat_interval'] as int? ?? 1,
      repeatEndDate: json['repeat_end_date'] != null
          ? DateTime.parse(json['repeat_end_date'] as String)
          : null,
      isCompleted: json['is_completed'] as bool? ?? false,
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  static RepeatType _parseRepeatType(String? type) {
    switch (type) {
      case 'hourly':
        return RepeatType.hourly;
      case 'daily':
        return RepeatType.daily;
      case 'weekly':
        return RepeatType.weekly;
      case 'monthly':
        return RepeatType.monthly;
      default:
        return RepeatType.none;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'category_id': categoryId,
      'category_name': categoryName,
      'title': title,
      'deadline': deadline?.toIso8601String(),
      'reminder_minutes': reminderMinutes,
      'repeat_type': repeatType.name,
      'repeat_interval': repeatInterval,
      'repeat_end_date': repeatEndDate?.toIso8601String(),
      'is_completed': isCompleted,
      'completed_at': completedAt?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Task copyWith({
    String? id,
    String? userId,
    String? categoryId,
    String? categoryName,
    String? title,
    DateTime? deadline,
    int? reminderMinutes,
    RepeatType? repeatType,
    int? repeatInterval,
    DateTime? repeatEndDate,
    bool? isCompleted,
    DateTime? completedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Task(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      categoryId: categoryId ?? this.categoryId,
      categoryName: categoryName ?? this.categoryName,
      title: title ?? this.title,
      deadline: deadline ?? this.deadline,
      reminderMinutes: reminderMinutes ?? this.reminderMinutes,
      repeatType: repeatType ?? this.repeatType,
      repeatInterval: repeatInterval ?? this.repeatInterval,
      repeatEndDate: repeatEndDate ?? this.repeatEndDate,
      isCompleted: isCompleted ?? this.isCompleted,
      completedAt: completedAt ?? this.completedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: server/internal/models/task.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type RepeatType string

const (
	RepeatNone    RepeatType = "none"
	RepeatHourly  RepeatType = "hourly"
	RepeatDaily   RepeatType = "daily"
	RepeatWeekly  RepeatType = "weekly"
	RepeatMonthly RepeatType = "monthly"
)

type Task struct {
	ID              string     `gorm:"type:varchar(36);primaryKey" json:"id"`
	UserID          string     `gorm:"type:varchar(36);not null;index:idx_user_id" json:"user_id"`
	CategoryID      *string    `gorm:"type:varchar(36);index:idx_category_id" json:"category_id"`
	Title           string     `gorm:"type:varchar(255);not null" json:"title"`
	Description     *string    `gorm:"type:text" json:"description,omitempty"`
	Deadline        *time.Time `json:"deadline,omitempty"`
	ReminderMinutes *int       `json:"reminder_minutes,omitempty"`
	DurationMinutes *int       `json:"duration_minutes,omitempty"`
	Difficulty      *string    `gorm:"type:varchar(20)" json:"difficulty,omitempty"` // relaxed, normal, focus
	RepeatType      RepeatType `gorm:"type:enum('none','hourly','daily','weekly','monthly');default:'none'" json:"repeat_type"`
	RepeatInterval  int        `gorm:"default:1" json:"repeat_interval"`
	RepeatEndDate   *time.Time `gorm:"type:date" json:"repeat_end_date,omitempty"`
	IsCompleted     bool       `gorm:"default:false;index:idx_is_completed" json:"is_completed"`
	CompletedAt     *time.Time `json:"completed_at,omitempty"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`

	// Relations
	User     User      `gorm:"foreignKey:UserID" json:"-"`
	Category *Category `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
}

// BeforeCreate hook untuk generate UUID
func (t *Task) BeforeCreate(tx *gorm.DB) error {
	if t.ID == "" {
		t.ID = uuid.New().String()
	}
	return nil
}

// TaskWithCategory response dengan nama kategori
type TaskWithCategory struct {
	Task
	CategoryName string `json:"category_name,omitempty"`
}
